### PR TITLE
[[ Bug 19911 ]] Ensure android interface callbacks are on engine thread

### PR DESCRIPTION
**Pull request by livecodeali**
Also document thread affinity syntax in binding string.

### CODE REVIEW

File: engine/src/mblandroidlcb.cpp [hunk 1]
```diff
@@ -55,13 +55,13 @@ extern "C" JNIEXPORT void JNICALL Java_com_runrev_android_LCBInvocationHandler_d
...
-    extern bool MCAndroidIsOnSystemThread(void);
-    if (!MCAndroidIsOnSystemThread())
+    extern bool MCAndroidIsOnEngineThread(void);
+    if (!MCAndroidIsOnEngineThread())
     {
         typedef void (*co_yield_callback_t)(void *);
-        extern void co_yield_to_android_and_call(co_yield_callback_t callback, void *context);
+        extern void co_yield_to_engine_and_call(co_yield_callback_t callback, void *context);
         remote_call_t t_context = {p_handler, p_method_name, p_args};
-        co_yield_to_android_and_call(remote_call_func, &t_context);
```
Comment (runrevmark) [inline engine/src/mblandroidlcb.cpp:64]: Isn't there a problem here with local JNI refs? The easiest way to fix this for now is to take GlobalRefs before doing the yield, and then delete them after the yield.

File: docs/guides/LiveCode Builder Language Reference.md [hunk 2]
```diff
@@ -554,8 +554,11 @@ non-Windows platform then it is taken to be `default`.
 
 The Java binding string has the following form:
 
-    "java:[className>][functionType.]function[!calling]"
-    
+    "java[thread]:[className>][functionType.]function[!calling]"
```
Comment (runrevmark) [inline docs/guides/LiveCode Builder Language Reference.md]: Is it worth considering putting the thread affinity somewhere else? Perhaps a final ?<thread-affinity> clause at the end.

Comment (livecodeali) [reply docs/guides/LiveCode Builder Language Reference.md]: Yeah, I guess it probably is

File: engine/src/mblandroidlcb.cpp (hunk 1 above)
Comment (livecodeali) [reply engine/src/mblandroidlcb.cpp:64]: Ah yes, I see what you mean

**Comment by livecodeali** [comment]
Depends on https://github.com/livecode/livecode/pull/5660 - will need rebasing once that is merged.

**Comment by runrevmark** [comment]
@livecode-vulcan review ok 73fb7f1

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 73fb7f1dacd37b1a5eb736271f6e73cdf90232e1

**Comment by livecode-vulcan** [comment]
😞 test failure 73fb7f1dacd37b1a5eb736271f6e73cdf90232e1

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/446)
* try-community-armv6-android-api9: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/454)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/433)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/376)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/376)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/441)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/446)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/444)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/454)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/444)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/453)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/471)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/440)

**Comment by livecodeali** [comment]
@runrevmark Fixed that build error.

**Comment by runrevmark** [comment]
@livecode-vulcan review ok e4fc525

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok e4fc5256f31d86b8b88fc6961b3ae59ad29c8fde

**Comment by livecode-vulcan** [comment]
😎 test success e4fc5256f31d86b8b88fc6961b3ae59ad29c8fde

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/450)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/458)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/437)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/380)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/380)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/445)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/450)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/448)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/461)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/448)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/457)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/475)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/444)

**Pull request merged**